bob-access/src/credentials.rs

```diff
@@ -9,18 +9,71 @@ use tonic::Request;
 
 use crate::{error::Error, extractor::ExtractorExt, Authenticator};
 
-#[derive(Debug, Default, Clone)]
-pub struct Credentials {
+#[derive(Debug, Default)]
+pub struct RequestCredentials {
     address: Option<SocketAddr>,
     kind: Option<CredentialsKind>,
 }
 
+impl RequestCredentials {
+    pub fn builder() -> RequestCredentialsBuilder {
+        RequestCredentialsBuilder::default()
+    }
+
+    pub fn kind(&self) -> Option<&CredentialsKind> {
+        self.kind.as_ref()
+    }
+
+    pub fn ip(&self) -> Option<SocketAddr> {
+        self.address
+    }
+}
+
+const RESOLVE_THRESHOLD: u8 = 3;
+
+#[derive(Debug, Clone)]
+enum ResolveState {
+    Waiting(u8),
+    Pending,
+}
+
+impl Default for ResolveState {
+    fn default() -> Self {
+        ResolveState::Waiting(0)
+    }
+}
+
+impl ResolveState {
+    fn needs_resolve(&mut self, unresolved: bool) -> bool {
+        match self {
+            ResolveState::Waiting(n) => {
+                if unresolved {
+                    *n += 1;
+                    *n >= RESOLVE_THRESHOLD
+                } else {
+                    *n = 0;
+                    false
+                }                
+            },
+            ResolveState::Pending => false,
+        }
+    }
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct DeclaredCredentials {
+    address: Option<Vec<SocketAddr>>,
+    hostname: Option<String>,
+    kind: Option<CredentialsKind>,
+    resolve_state: ResolveState,
```

Comment:
`DeclaredCredentials` is not a good place for `ResolveState`. It should contain only fields, required for authentication. It is better to create a wrapper struct, that will contain both  `DeclaredCredentials` and  `ResolveState`